src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,409 @@
+		return "", fmt.Errorf("Couldn't parse cluster.yaml. Error: %w.", err)
+	}
+
+	// As of 2.0.2, the following microcluster method will:
+	// * validate the member changes
+	//     * ensure that no members were added or removed
+	//     * the member IDs hasn't changed
+	//     * there is at least one voter
+	//     * the addresses can be parsed
+	//     * there are no duplicate addresses
+	// * ensure that all cluster members are stopped
+	// * create a database backup
+	// * reconfigure Raft
+	// * address changes based on the new cluster.yaml:
+	//     * refresh the local info.yaml and daemon.yaml
+	//     * update the trust store addresses
+	//     * prepare an sql script that updates the member addresses from the
+	//       "core_cluster_members" table, executed when k8sd starts.
+	// * rewrite cluster.yaml
+	// * create a recovery tarball of the k8sd database dir and store it
+	//   in the state dir.
+	tarballPath, err := m.RecoverFromQuorumLoss(newMembers)
+	if err != nil {
+		return "", fmt.Errorf("K8sd recovery failed: %w", err)
+	}
+
+	return tarballPath, nil;
+}
+
+func recoverK8sDqlite() (string, string, error) {
+	k8sDqliteStateDir := clusterRecoverOpts.K8sDqliteStateDir
+
+	clusterYamlPath := path.Join(k8sDqliteStateDir, "cluster.yaml")
+	oldMembersYaml, err := os.ReadFile(clusterYamlPath)
+	if err != nil {
+		return "", "", fmt.Errorf("Could not read k8s-dqlite cluster.yaml: %w.", err);
+	}
+	clusterYamlCommentHeader := fmt.Sprintf("# k8s-dqlite cluster configuration\n# (%s)\n", clusterYamlPath)
+	clusterYamlContent, err := shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(clusterYamlCommentHeader),
+			[]byte(clusterYamlRecoveryComment),
+			[]byte(yamlHelperCommentFooter),
+			oldMembersYaml))
+	if err != nil {
+		return "", "", fmt.Errorf("Interactive text editor failed: %w.", err);
+	}
+	if err = removeYamlHelperCommentsFromFile(clusterYamlPath); err != nil {
+		return "", "", fmt.Errorf("Couln't remove guide comments: %s. Error: %s.", clusterYamlPath, err)
+	}
+
+	infoYamlPath := path.Join(k8sDqliteStateDir, "info.yaml")
+	infoYamlCommentHeader := fmt.Sprintf("#k8s-dqlite info.yaml\n# (%s)\n", infoYamlPath)
+	_, err = shared.TextEditor(
+		"",
+		slices.Concat(
+			[]byte(infoYamlCommentHeader),
+			[]byte(infoYamlRecoveryComment),
+			[]byte(yamlHelperCommentFooter),
+			utils.YamlCommentLines(removeYamlHelperComments(clusterYamlContent))))
+	if err != nil {
+		return "", "", fmt.Errorf("Interactive text editor failed: %w.", err);
+	}
+	if err = removeYamlHelperCommentsFromFile(infoYamlPath); err != nil {
+		return "", "", fmt.Errorf("Couln't remove guide comments: %s. Error: %s.", infoYamlPath, err)
+	}
+
+	newMembers := []cluster.DqliteMember{}
+	if err = yaml.Unmarshal(clusterYamlContent, &newMembers); err != nil {
+		return "", "", fmt.Errorf("Couldn't parse cluster.yaml. Error: %w.", err)
+	}
+
+	if err = ensureK8sDqliteMembersStopped(newMembers); err != nil {
```

Comment:
This check is done quite late in the recovery process (e.g. after the user edited the configuration). If this fails, then the user did useless work. 

You could move this at the top of `recoverK8sDqlite` but I would even put it into `recoveryCmdPreflightChecks` to ensure that we are not failing half-way through the process.